### PR TITLE
pkg/memio - Rework&Cover

### PR DESCRIPTION
**Pull request by ChriMarMe**
* Add interface implementation for platform independent MemIO read and write
* Add unittests in table form for MemIO interface and implementation
* Add interface Port as first abstraction for MemIO interface
* Adapt linux and plan9 implementation to new interface. Caller can use it without knowledge of the platform
* Add tests in table form for linux implementation
* Add tests for UintN type
* Add syscall abstraction and ReadAt, WriteAt to mmap linux implentation and adapt tests to table form

Signed-off-by: Christopher Meis <christopher.meis@9elements.com>

### CODE REVIEW

**Comment by ChriMarMe** [comment]
Tough to please the CI these days.

File: pkg/memio/memio_ports.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+type MemIOReader interface {
+	Read(UintN, int64) error
+}
+
+type MemIOWriter interface {
+	Write(UintN, int64) error
+}
+
+type MemIO interface {
```
Comment (rminnich) [inline pkg/memio/memio_ports.go:23]: The convention I think, albeit awkward, might be MemIOReadWriteCloser

File: pkg/memio/memio_ports.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	if _, err := m.File.Seek(addr, io.SeekStart); err != nil {
+		return err
+	}
+	return binary.Write(m.File, ubinary.NativeEndian, in)
+}
+
+func (m *MemIOPort) Close() error {
+	return m.File.Close()
+}
+
```
Comment (rminnich) [inline pkg/memio/memio_ports.go:50]: var _ MemIO = &MemIOPort{}

File: pkg/memio/ports_linux.go [hunk 3]
```diff
@@ -9,86 +9,68 @@ package memio
...
-	switch p := data.(type) {
-	case *Uint32:
-		*p = Uint32(archInl(addr))
-	case *Uint16:
-		*p = Uint16(archInw(addr))
-	case *Uint8:
-		*p = Uint8(archInb(addr))
-	default:
-		return fmt.Errorf("port data must be 8, 16 or 32 bits")
+// NewPort returns a new instance of LinuxPort for read/write operations on /dev/port
```
Comment (rminnich) [inline pkg/memio/ports_linux.go:46]: Also add the var _ etc. etc. type assertion here.

**Review by rminnich — COMMENTED** [review]
This is a substantial improvement. See what you think of my nits, the one thing I thing owuld be good is the type assertion.

File: pkg/memio/memio_ports.go (hunk 2 above)
Comment (ChriMarMe) [reply pkg/memio/memio_ports.go:50]: I'm pretty sure type assertion looks like this:
`unknownType.(concreteType)`

Do you have a ref doc for your suggested construct of assertion?

Comment (ChriMarMe) [reply pkg/memio/memio_ports.go:50]: Ah, never mind.

**Comment by rjoleary** [comment]
Does this come before or after https://github.com/u-root/u-root/pull/2190 ? It looks like we'll have a merge conflict.

**Comment by ChriMarMe** [comment]
We forgot to close #2190 , but now it' closed.

**Review by walterchris — APPROVED** [review]

**Pull request merged**